File: src/main/java/io/xol/chunkstories/api/plugin/ChunkStoriesPlugin.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.plugin;

import java.io.File;

import io.xol.chunkstories.api.GameContext;

public abstract class ChunkStoriesPlugin {
	protected final GameContext pluginExecutionContext;

	private final PluginInformation pluginInformation;

	public ChunkStoriesPlugin(PluginInformation pluginInformation, GameContext pluginExecutionContext) {
		this.pluginInformation = pluginInformation;
		this.pluginExecutionContext = pluginExecutionContext;
	}

	public PluginInformation getPluginInformation() {
		return pluginInformation;
	}

	public GameContext getPluginExecutionContext() {
		return pluginExecutionContext;
	}

	public PluginManager getPluginManager() {
		return pluginExecutionContext.getPluginManager();
	}
	
	public File getDirectory() {
		return pluginInformation.getDirectory();
	}

	public abstract void onEnable();

	public abstract void onDisable();

	public String getName() {
		return pluginInformation.getName();
	}
}

File: src/main/java/xyz/chunkstories/api/plugin/ChunkStoriesPlugin.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.plugin

import java.io.File

import xyz.chunkstories.api.GameContext

abstract class ChunkStoriesPlugin(val pluginInformation: PluginInformation, val gameContext: GameContext) {

    val pluginManager: PluginManager
        get() = gameContext.pluginManager

    val directory: File = pluginManager.getPluginDirectory(this)

    val name: String
        get() = pluginInformation.name

    abstract fun onEnable()

    abstract fun onDisable()
}
